// mysql-nodejs/lib/errors.js
const util = require('util');

function InvalidParameter( message ){
    this.name = "Invalid Parameter Error";
    this.message = message || "Please check your parameter provided";
    Error.captureStackTrace( this, InvalidParameter );
};

util.inherits( InvalidParameter, Error );

function YetToConnectError(){
    this.name = "Yet To Connect Error";
    this.message = "Please create connection before use";
    Error.captureStackTrace( this, YetToConnectError );
};

util.inherits( YetToConnectError, Error );

function AlreadyConnectError(){
    this.name = "Already Connect Error";
    this.message = "Already created connection";
    Error.captureStackTrace( this, AlreadyConnectError );
};

util.inherits( AlreadyConnectError, Error );

function MethodNotAllow( message ){
    this.name = "Method not allow";
    this.message = message || "Please check your methods";
    Error.captureStackTrace( this, MethodNotAllow );
};

util.inherits( MethodNotAllow, Error );

module.exports = {
    InvalidParameter: InvalidParameter,
    YetToConnectError: YetToConnectError,
    AlreadyConnectError: AlreadyConnectError,
    MethodNotAllow: MethodNotAllow
};
